lib/Common/Memory/SmallFinalizableHeapBucket.h

```diff
@@ -174,7 +173,7 @@ class HeapBucketGroup
         typedef typename SmallHeapBlockType<objectAttributes, TBlockAttributes>::BucketType BucketType;
         static BucketType& GetBucket(HeapBucketGroup<TBlockAttributes> * heapBucketGroup)
         {
-            CompileAssert(objectAttributes & FinalizeBit);
+            Assert(objectAttributes & FinalizeBit);
```

Comment:
we are not actually calling, with following:

#if GLOBAL_ENABLE_WRITE_BARRIER
    if (CONFIG_FLAG(ForceSoftwareWriteBarrier)
        && ((attributes & LeafBit) != LeafBit
            || (attributes & FinalizeBit) == FinalizeBit)) // there's no Finalize Leaf bucket
    {
        memBlock = RealAlloc<(ObjectInfoBits)((attributes | WithBarrierBit) & InternalObjectInfoBitMask), nothrow>(&autoHeap, allocSize);
    }
    else
#endif
    {
        memBlock = RealAlloc<(ObjectInfoBits)(attributes & InternalObjectInfoBitMask), nothrow>(&autoHeap, allocSize);
    }


the LeafBit without FinalizeBit has already going through the path with LeafBit specialization and won't fall through here. when compiler doing optimization, it should remove such symbols. even if it's not, the runtime assertion should catch it.
just because the compiler didn't analyze the 'if' statement while expanding the template. 

more clear way to explain this is, compiler thought there's (WithBarrierBit | LeafBit) combination (deducting from (attributes | WithBarrierBit)), which won't hit the LeafBit specialization and fall through the default one which we Assertion on FinalizeBit is set.
however we actually don't have (WithBarrierBit | LeafBit) combination, from the if statement, if the original attribute is LeafBit only, I don't adding WithBarrierBit even with ForceSoftwareWriteBarrier on


---
In reply to: [90006829](https://github.com/Microsoft/ChakraCore/pull/2105#discussion_r90006829) [](ancestors = 90006829)